Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -95,6 +143,106 @@ private static void Warn(DafnyOptions dafnyOptions, ErrorReporter reporter, Proo
+      SuggestFunctionHiding(unusedFunctions, m);
+      SuggestByProofRefactoring(scopeName, assertCoverage.ToList());
+    }
+  }
+
+  private static void SuggestFunctionHiding(List<Function> unusedFunctions, Method m) {
+    if (unusedFunctions.Any()) {
+      reporter.Info(MessageSource.Verifier, m.Body.StartToken,
+        $"Consider hiding {(unusedFunctions.Count > 1 ? "these functions, which are" : "this function, which is")} unused by the proof: {unusedFunctions.Comma()}");
+    }
+  }
+
+  private static void SuggestByProofRefactoring(string scopeName,
+    IReadOnlyList<DafnyConsolePrinter.VerificationRunResultPartialCopy> verificationRunResults) {
+    foreach (var (dep, lAsserts) in ComputeAssertionsUsedByFact(scopeName, verificationRunResults)) {
+      var factIsOnlyUsedByOneAssertion = lAsserts.Count == 1;
+      if (!factIsOnlyUsedByOneAssertion) {
+        continue;
+      }
+
+      DafnyConsolePrinter.AssertCmdPartialCopy cmd = null;
+      foreach (var assert in lAsserts) {
+        cmd = assert;
+      }
+
+      manager.ProofDependenciesById.TryGetValue(cmd.Id, out var consumer);
+
+      if (consumer != null && (dep == consumer || consumer.Range.Intersects(dep.Range))) {
+        continue;
+      }
+
+      RangeToken range = null;
+      var factProvider = "";
+      var factConsumer = "";
+      var recommendation = "";
+      var completeInformation = true;
+
+      switch (dep) {
+        case AssumedProofObligationDependency:
+        case AssumptionDependency: {
+            range = dep.Range;
+            factProvider = "fact";
+            recommendation = "moving it into";
+            break;
+          }
+        case RequiresDependency: {
+            range = dep.Range;
+            factProvider = "requires clause";
+            recommendation = "labelling it and revealing it in";
+            break;
+          }
+        default: completeInformation = false; break;
+      }
+
+      switch (consumer) {
+        case CallDependency call: {
+            factConsumer = $"precondtion{(call.call.Method.Req.Count > 1 ? "s" : "")} of the method call {call.RangeString()}";
+            break;
+          }
+        case ProofObligationDependency { ProofObligation: AssertStatementDescription }: {
+            factConsumer = $"assertion {consumer.RangeString()}";
+            break;
+          }
+        default: completeInformation = false; break;
+      }
+
+      if (completeInformation) {
+        reporter.Info(MessageSource.Verifier, range,
+          $"This {factProvider} was only used to prove the {factConsumer}. Consider {recommendation} a by-proof.");
+      }
+    }
+  }
+
+  private static Dictionary<ProofDependency, HashSet<DafnyConsolePrinter.AssertCmdPartialCopy>>
+    ComputeAssertionsUsedByFact(string scopeName, IReadOnlyList<DafnyConsolePrinter.VerificationRunResultPartialCopy> vcResults) {
+    var assertionsUsedByFact = manager.GetPotentialDependenciesForDefinition(scopeName)
```

Comment:
Actually.. shouldn't `assertionsUsedByFact` be called `assertionsProvenUsingFact` ?